src/main/software/amazon/event/ruler/NameState.java

```diff
@@ -170,7 +280,41 @@ public String toString() {
         return "NameState{" +
                 "valueTransitions=" + valueTransitions +
                 ", mustNotExistMatchers=" + mustNotExistMatchers +
-                ", rules=" + rules +
+                ", patternToTerminalSubRules=" + patternToTerminalSubRules +
+                ", patternToNonTerminalSubRules=" + patternToNonTerminalSubRules +
                 '}';
     }
-}
+
+    /**
+     * Store an ID with a rule to represent a sub-rule, which differentiates between rules of the same name.
+     */
+    static class SubRule {
```

Comment:
It sounds good to assign a unique Id for each sub rule, just the sub rule can have big portion of overlap, e.g.
```
rule: |-sub_r1:"a":"a1"                                                       
      |-sub_r2:"a":"a1"->"b":"b1"
      |-sub_r3:"a":"a1"->"b":"b1"->"c":"c1
```
ideally, it would be great to save more if subrule object can be reused for the overlapped field/values along the path, so we can limit the size of `Set<SubRule>` to only name state where the new branch starts.
assume each field have a unique id,
in above example, the id in N1, N2, N3 name state would be:
Id(r1) = id1
Id(r2) = Id(r1)_id2 =  id1_id2
Id(r3) = Id(r2)_id3 = Id1_id2_id3
the  `Set<SubRule>` for given pattern will only be 1,instead of 3 for each sub rule in each its name states.

I see it is hard to model it with high perf, the alternative is to not track the rule, but try `List<String, List<Patterns>> namevals` and `keyIndex` ( field index),  which patterns of value at `keyIndex` field doesn't matter actually, so the same subrule set even can be reused cross rules)

just my 2 cents for your reference.

Reply:
Note that "Object rule" inside SubRule is just the name, not the full rule body. I'm not sure I follow your suggestion entirely. Even making the SubRule recursive in some way would still result in the same total number of SubRules. But I'm pretty sure the suggestion assumes that "Object rule" is the full rule body.

Reply:
yes, I think we actually care more about the rule body than rule name as the "path" - I like call it "footprint" is relevant the rule body.
I way trying to explain a way to reduce the size of Set<SubRule> by reusing the subrule id if sub rule is an extension of exiting rule.
current way:
N1
sub rule id 1
sub rule id 2
sub rule id 3
N2
sub rule id 1
sub rule id 2
sub rule id 3
N3
sub rule id 1
sub rule id 2
sub rule id 3

My mental model with the proposal is:
N1
sub rule id 1
N2
sub rule id 1-2
N3
sub rule id 1-2-3
The first portion "1" reflects rule name, sub rules with starting of "1" means they are for the same rule, the sub rule 2 is an extension from sub-rule 1, so its id could also be extension from sub rule id 1 - "1-2" ... similar for next one ...

just share for your reference, I see it seems overoptimizing.